File: 039.rb
# 039
# Combination Sum
# 2015-05-16
#####################################################
# @param {Integer[]} candidates
# @param {Integer} target
# @return {Integer[][]}
def combination_sum(candidates, target)
  candidates.sort!()
  solve(candidates, target, candidates.size-1)
end

def solve(candidates, target, limit)
  v = []
  return v if target < candidates.first
  v.push([target]) if candidates[0..limit].include?(target)
  for i in (0..limit)
    break if candidates[i] == target
    solve(candidates, target-candidates[i], i).each do |a|
      a.push(candidates[i])
      v.push(a)
    end
  end
  v
end

#####################################################

candidates = [2, 3, 6, 7]
target = 7

combination_sum(candidates, target).each do |a|
  puts a.inspect
end

File: python/039.py
#   039
#   Combination Sum
#   2015-05-16
#####################################################
class Solution:
    # @param {integer[]} candidates
    # @param {integer} target
    # @return {integer[][]}
    def combinationSum(self, candidates, target):
        candidates.sort()
        return self.solve(candidates, target, len(candidates)-1)

    def solve(self, candidates, target, limit):
        v = []
        if target < candidates[0]:
            return v
        if target in candidates[:limit+1]:
            v.append([target])
        for i in range(limit+1):
            if target == candidates[i]:
                break
            for a in self.solve(candidates, target-candidates[i], i):
                a.append(candidates[i])
                v.append(a)
        return v

#####################################################

candidates = [3, 2, 1]
target = 6

sol = Solution()
for a in sol.combinationSum(candidates, target):
    print(a)

